tensorflow/python/ops/special_math_ops.py

```diff
@@ -94,48 +94,100 @@ def einsum(axes, *inputs):
+                or the number of dimensions of an input differs from the number of indices in its subscript,
+                or the input shapes are inconsistent along a particular axis.
   """
+  if '...' in axes:
+    raise ValueError("Subscripts with ellipses are not yet supported.")
 
-  match = re.match('([a-z,]+)->([a-z]+)', axes)
-  assert match, \
-    "Indices have incorrect format: %s" % axes
+  match = re.match('([a-z,]+)(->[a-z]*)?', axes)
+  if not match:
+    raise ValueError(
+      "Indices have incorrect format: %s" % axes
+    )
 
   inputs = list(inputs)
   idx_in = match.group(1).split(',')
-  idx_out = match.group(2)
   idx_all = set(''.join(idx_in))
+  indices = ''.join(sorted(idx_all))
+
+  if match.group(2):
+    idx_out = match.group(2)[2:]
 
+  else:
+    # infer the output subscripts if not given, assume alphabetical order
+    counts = {ax: 0 for ax in indices}
+    for axes_ in idx_in:
+      for ax in axes_:
+        counts[ax] += 1
 
-  assert len(idx_in) == len(inputs), \
-    "Expected %d inputs but only got %d" % (len(idx_in), len(inputs))
+    idx_out = ''.join(sorted(
+      ax for ax in indices
+      if counts[ax] == 1
+    ))
 
-  # transpose inputs so axes are in alphabetical order
+  if len(idx_in) != len(inputs):
+    raise ValueError(
+      "Expected %d inputs but got %d" % (len(idx_in), len(inputs))
+    )
+
+  missing_idx = set(idx_out).difference(idx_all)
+  if missing_idx:
+    raise ValueError(
+      "Unknown ouput axes: %s" % missing_idx
+    )
+
+  axis_order = {}
+  for ax in indices:
+    if ax not in idx_out:
+      axis_order[ax] = len(axis_order)
+  for ax in idx_out:
+    axis_order[ax] = len(axis_order)
+
+  # transpose inputs so axes are in order
   for i, (input_, axes_) in enumerate(zip(inputs, idx_in)):
-    assert input_.get_shape().ndims == len(axes_), \
-      "Input %d with axes %s has incorrect" \
-      " number of dimensions (expected %d, got %d)" % (
-        i, axes_, len(axes_), input_.get_shape().ndims
+    if input_.get_shape().ndims != len(axes_):
+      raise ValueError(
+        "Input %d with axes %s has incorrect" \
+        " number of dimensions (expected %d, got %d)" % (
+          i, axes_, len(axes_), input_.get_shape().ndims
+        )
       )
 
-    sorted_idx = sorted(axes_)
+    sorted_idx = sorted(axes_, key=axis_order.get)
+
+    if len(set(axes_)) != len(axes_):
+      raise ValueError(
+        "Subscript not supported: an axis appears more than once: %s" % axes_
+      )
 
     if list(axes_) != sorted_idx:
       permuted = [axes_.find(ax) for ax in sorted_idx]
       inputs[i] = array_ops.transpose(input_, permuted)
       idx_in[i] = sorted_idx
 
-  missing_idx = set(idx_out).difference(idx_all)
-  assert not missing_idx, \
-    "Unknown ouput axes: %s" % missing_idx
-
   reduction_idx = []
   shapes = [[dim if dim else -1
              for dim in tensor.get_shape().as_list()]
```

Comment:
I would like to propose one more fix. I think this should be changed from:

```
shapes = [[dim if dim else -1
           for dim in tensor.get_shape().as_list()]
           for tensor in inputs]
```

to:

```
shapes = [[dim if dim is not None else -1
           for dim in array_ops.unpack(array_ops.shape(tensor))]
           for tensor in inputs]
```

in order to allow for dynamic tensor shapes. I run into an issue while using `tf.einsum` and this change fixed it for me.